### ISSUE TITLE
bugprone-implicit-widening-of-multiplication-result improvements

### ISSUE THREAD

**Issue by chfast** [llvm/llvm-project#52806]
This is nice check, but it is also triggered for a multiplication of only constants.
```
long x = 1 * 1;
long y = 4 * 1024;
```
I think the check should ignore such situation. I assume there is a compiler warning if the constant multiplication result is out-of-range of the destination type.

Moreover, the suggested warning suppression is `static_cast<long>(4 * 1024)`. However a variant of this `long{4 * 1024}` does not suppress the warning.

I believe the better way would be to use the actual type of the multiplication: `int{4 * 1024}`. This actually works as the warning suppression.

https://godbolt.org/z/Y1xYn988e

**Comment by PeterFeicht** [comment]
Partial duplicate of #58709.

**Comment by cwarner-8702** [comment]
A PR was merged yesterday that should suppress warnings for signed integer types when the result of the expression can be calculated at compile time (unsigned integers are a little more complicated, so it may take a bit longer): https://github.com/llvm/llvm-project/commit/f964e8a3f9ef9789f9075694cb887d0b3986d304

**Comment by vbvictor** [comment]
No longer reproductible on trunk with option
`-config='{CheckOptions: {bugprone-implicit-widening-of-multiplication-result.IgnoreConstantIntExpr: true}}'`
https://godbolt.org/z/exaGE5bcK
